src/main/scala/uk/gov/ons/registers/methods/PayeCalculator.scala

```diff
@@ -0,0 +1,152 @@
+    val calculatedPayeJobsDF = getGroupedByPayeJobs(BIDF, payeDF, "dec_jobs")
+    val calculatedPayeDF = calculatedPayeEmployeesDF.join(calculatedPayeJobsDF, "ern")
+    //calculatedPayeDF.show
+
+    calculatedPayeDF
+  }
+
+  def getGroupedByPayeEmployees(BIDF: DataFrame, payeDF: DataFrame, luTableName: String = "LEGAL_UNITS", payeDataTableName: String = "PAYE_DATA")(implicit spark: SparkSession): DataFrame ={
+    val flatUnitDf = BIDF.withColumn("payeref", explode_outer(BIDF.apply("PayeRefs")))
+    val idDF1 = (flatUnitDf.join(payeDF, "payeref"))
+    val idDF = idDF1.selectExpr("ern", "id", "cast(mar_jobs as int) mar_jobs", "cast(june_jobs as int) june_jobs", "cast(sept_jobs as int) sept_jobs", "cast(dec_jobs as int) dec_jobs")
+      .groupBy("id").agg(sum("mar_jobs") as "mar_jobs", sum("june_jobs") as "june_jobs", sum("sept_jobs") as "sept_jobs", sum("dec_jobs") as "dec_jobs")
+
+    missingPayeRefsThrow(flatUnitDf,idDF1)
+
+    idDF.createOrReplaceTempView(payeDataTableName)
+    flatUnitDf.createOrReplaceTempView(luTableName)
+
+    val flatPayeDataSumSql = generateCalculateSumSQL(luTableName, payeDataTableName)
+    val flatPayeDataCountSql = generateCalculateCountSQL(luTableName, payeDataTableName)
+
+    val sqlSum =  s"""
+              SELECT (SUM(AVG_CALCULATED.quarter_sum)) AS sums, AVG_CALCULATED.id, AVG_CALCULATED.ern
+              FROM ($flatPayeDataSumSql) as AVG_CALCULATED
+              GROUP BY AVG_CALCULATED.ern, AVG_CALCULATED.id
+            """.stripMargin
+    val sqlCount = s"""
+              SELECT (SUM(AVG_CALCULATED.quarter_count)) AS counts, AVG_CALCULATED.id
+              FROM ($flatPayeDataCountSql) as AVG_CALCULATED
+              GROUP BY AVG_CALCULATED.id
+            """.stripMargin
+    val Sum = spark.sql(sqlSum)
+    val Count = spark.sql(sqlCount)
+    val aggDF = Sum.join(Count, "id")
+
+    val ungroupedDF = aggDF.withColumn(employees, aggDF.col("sums") / aggDF.col("counts"))//.selectExpr("ern", s"cast($employees as int) $employees")
+    val groupedDF = ungroupedDF.groupBy("ern").agg(sum(employees) as employees).selectExpr(s"cast($employees as int) $employees", "ern")
+
+    groupedDF
+  }
+
+  def getGroupedByPayeJobs(BIDF: DataFrame, payeDF: DataFrame, quarter: String,luTableName: String = "LEGAL_UNITS", payeDataTableName: String = "PAYE_DATA")(implicit spark: SparkSession): DataFrame ={
+    val flatUnitDf = BIDF.withColumn("payeref", explode_outer(BIDF.apply("PayeRefs")))
+    val idDF = (payeDF.join(flatUnitDf, "payeref")).selectExpr("ern", s"cast($quarter as int) $quarter").groupBy("ern").agg(sum(quarter) as jobs)
+    idDF
+  }
+
+  def generateCalculateSumSQL(luTablename: String = "LEGAL_UNITS", payeDataTableName: String = "PAYE_DATA") =
+    s"""
+      SELECT $luTablename.*, $payeDataTableName.mar_jobs, $payeDataTableName.june_jobs, $payeDataTableName.sept_jobs, $payeDataTableName.dec_jobs,
+                CAST(
+               (
+                (CASE
```

Comment:
You are building a dynamic SQL string here.
You can still introduce helper functions to help you build the string - there is no need to define the null cast to zero in-line for each and every variable for example ...